### PR TITLE
Lab 05. Entity Framework

### PR DESCRIPTION
**Pull request by Zhunusau**
## Lab 05. Entity Framework

1. Changed Data Application Layer (DAL), used Code First approach with Entity Framework 6.2.
2. Added Fluent API Configurations in Entity Framework.
3. Added Data into database with a custom Initializer.
4. Added Migrations.
## Sonar
http://192.168.8.119:9000/dashboard?id=Ruslan-Zhunusau

### CODE REVIEW

File: Lab05.EF/Lab05.EF.BLL/Services/FileService.cs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                {
+                    unit.Repository<MetaFile>().Delete(metaFile);
+                    unit.Commit();
+                }
+                catch (DataException e)
+                {
+                    unit.Rollback();
+                    throw new DataException(e.Message);
+                }
+                unit.FileRepository.Delete(metaFile.Name);
```
Comment (a-yandulski) [inline Lab05.EF/Lab05.EF.BLL/Services/FileService.cs]: Deletion of the file must occur  inside `try...catch`, before unit.Commit(). If you can't  delete the file then you should not remove it's metadata from db

File: Lab05.EF/Lab05.EF.DAL/Context/StorageContext.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+using Lab05.EF.DAL.Interfaces;
+using Lab05.EF.DAL.Models;
+
+namespace Lab05.EF.DAL.Context
+{
+    public class StorageContext : DbContext, IStorageContext
+    {
+        public StorageContext() : base("StorageConnection")
+        {
+            Database.SetInitializer(new StorageInitializer());
```
Comment (a-yandulski) [inline Lab05.EF/Lab05.EF.DAL/Context/StorageContext.cs]: Would be nice to have database initialized on application startup. With current approach I'll have to perform operations with db to get it created\migrated. Example: https://stackoverflow.com/questions/12724320/forcing-code-first-to-always-initialize-a-non-existent-database
Please implement necessary service class for wrapping this functionality

File: Lab05.EF/Lab05.EF.DAL/Repositories/UnitOfWork.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        public void Commit()
+        {
+            dbContext.SaveChanges();
+        }
+
+        public void Rollback()
+        {
+            foreach (var entry in dbContext.ChangeTracker.Entries()
```
Comment (a-yandulski) [inline Lab05.EF/Lab05.EF.DAL/Repositories/UnitOfWork.cs]: I would move this code to the context class. Name of the method can be something like Reset().  So here you will just cal l: 
`dbContext.Reset();`

File: Lab05.EF/PluginJson/PluginJson.csproj [hunk 4]
```diff
@@ -0,0 +1,60 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (a-yandulski) [inline Lab05.EF/PluginJson/PluginJson.csproj]: Plugin projects should follow naming conventions: **Lab05.EF.**PluginJson.csproj. Rename json, xml and yaml plugin projects

File: Lab05.EF/Lab05.EF.BLL/Services/FileService.cs (hunk 1 above)
Comment (Zhunusau) [reply Lab05.EF/Lab05.EF.BLL/Services/FileService.cs]: fixed

File: Lab05.EF/Lab05.EF.DAL/Repositories/UnitOfWork.cs (hunk 3 above)
Comment (Zhunusau) [reply Lab05.EF/Lab05.EF.DAL/Repositories/UnitOfWork.cs]: moved to the StorageContext and renamed to Reset()

File: Lab05.EF/Lab05.EF.Config/Config.cs [hunk 5]
```diff
@@ -5,20 +5,20 @@ namespace Lab05.EF.Config
 {
     public class Config : IConfig
     {
-        private readonly IUserSection userSection;
+        //private readonly IUserSection userSection;
```
Comment (a-yandulski) [inline Lab05.EF/Lab05.EF.Config/Config.cs]: Remove comments

File: Lab05.EF/Lab05.EF/App.config [hunk 6]
```diff
@@ -1,7 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <configuration>
   <configSections>
-    <section name="userSection" type="Lab05.EF.Config.AppSettings.UserSection, Lab05.EF.Config" />
+    <!--<section name="userSection" type="Lab05.EF.Config.AppSettings.UserSection, Lab05.EF.Config" />-->
```
Comment (a-yandulski) [inline Lab05.EF/Lab05.EF/App.config]: Remove commented lines

File: Lab05.EF/Lab05.EF.BLL/Services/FileService.cs [hunk 7]
```diff
@@ -151,13 +151,14 @@ public void Remove(string fileName)
                 {
                     unit.Repository<MetaFile>().Delete(metaFile);
                     unit.Commit();
+                    unit.FileRepository.Delete(metaFile.Name);
```
Comment (a-yandulski) [inline Lab05.EF/Lab05.EF.BLL/Services/FileService.cs]: unit.FileRepository.Delete(metaFile.Name) must be called **before** unit.Commit()

File: Lab05.EF/Lab05.EF/App.config (hunk 6 above)
Comment (Zhunusau) [reply Lab05.EF/Lab05.EF/App.config]: removed comments

File: Lab05.EF/Lab05.EF.Config/Config.cs (hunk 5 above)
Comment (Zhunusau) [reply Lab05.EF/Lab05.EF.Config/Config.cs]: removed comments

File: Lab05.EF/Lab05.EF.BLL/Services/FileService.cs (hunk 7 above)
Comment (Zhunusau) [reply Lab05.EF/Lab05.EF.BLL/Services/FileService.cs]: fixed

**Pull request merged**